design/0003-byop.md

```diff
@@ -0,0 +1,176 @@
+re-use their exisitng prometheus.
+
+The following are the list of problems that we plan to solve:
+
+- Hard requirement of linkerd-prometheus.
+
+- Configuring the current linkerd-prometheus for remote-writes, etc is
+hard, because the prometheus configmap is overridden after upgrades.
+
+- Not enough documentation and tooling for bringing your own prometheus
+use-case.
+
+Once the RFC is implemented, Users should not only be able to use their
+exisiting prometheus, but configuration of the linkerd-prometheus for
+remote-writes, etc should also be easy.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### Prometheus as an Add-On
+
+By moving Prometheus as an add-on, We can make it as an optional component
+that can be used both through Helm and Linkerd CLI. The add-on model would
+well with upgrades, etc as per [#3955](https://github.com/linkerd/linkerd2/pull/3955).
+This is done by moving the prometheus components into a separate sub-chart
+in the linkerd repo and updating the code. The same health checks are also
+expected, but as a separate check Category, which would be transient i.e
+runs only when it recognises that the add-on is enabled.
+
+By default, the prometheus add-on will be **enabled**.
+
+This also provides a single place of configuration for all things related to
+linkerd-prometheus. Though we are making linkerd-prometheus optional but
+having it is great for Linkerd and external users, for reasons like
+separation of concerns, etc. By allowing the `remote-write`, `alerts`, etc
+configuration in the linkerd-prometheus, we would decrease the need/want
+for external users to tie their existing prometheus to Linkerd,
+as it would be even easier for them to configure `remote-write`, etc
+configuration from their exisiting prometheus to the linkerd-prometheus
+right from the start.
+
+The following default prometheus configuration is planned:
+
+```yaml
+prometheus:
+  enabled: true
+  name: linkerd-prometheus
+  image:
+  global:
+    <global-prom-config> https://prometheus.io/docs/prometheus/latest/configuration/configuration/#configuration-file
+  remote-write:
+    <remote-write-config> https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write
+  alerts:
+    <alert-manager-config> https://prometheus.io/docs/prometheus/latest/configuration/configuration/#alertmanager_config 
+```
+
+All the current default configuration will be moved to `values.yaml` in the
+above format, allowing users to override this by using Helm or  
+`--addon-config` through CLI.
+
+### Using Exisiting Prometheus
+
+First, Documentation will be provided at `linkerd.io` to guide users on
+adding the `scrape-configs` required, to get the metrics from linkerd
+proxies, control plane components, etc. This can be found
+[here](https://github.com/linkerd/linkerd2/blob/master/charts/linkerd2/templates/prometheus.yaml#L26).
+This can also be automated like we already do with Linkerd CLI help and args.
```

Comment:
I agree some automation via CLI is useful. Some kind of CLI command that outputs the Prometheus ConfigMap necessary to scrape Linkerd proxies. We don't change that scrape config often, but we'll want to ensure that the ConfigMap provided to users matches their Linkerd installation.

Additionally, we should have some documentation on the storage requirements for Linkerd metrics in Prometheus. We workaround this today by having a 6h retention time, but this could become a significant issue for existing Prometheus installations with longer retention times. This could motivate customizations to the scrape config (like scrape interval changes), to help decrease storage requirements.

Reply:
Added this. On the additional documentation part, I definitely think we have some stuff in mind around retention times, etc, and more can be gathered as we understand problems when users use their existing prometheus.